File: Maximum Number of coins.cpp
class Solution{
    public:
        int maxCoins(int N, vector <int> &a)
        {
            a.push_back(1);
            a.insert(a.begin(), 1);
            vector<vector<int> > dp(a.size(), vector<int>(a.size(), 0));
            for (int i = a.size() - 3; i >= 0; i --) {
                for (int j = i + 2; j < a.size();j ++) {
                    for (int k = i + 1; k < j; k ++)
                        dp[i][j] = max(dp[i][j], dp[i][k] + dp[k][j] + a[i] * a[k] * a[j]);
                }
            }
            return dp[0][a.size() - 1];    
        }
};

File: Maximum Number of coins.py
class Solution():
    def maxCoins(self, N, a):
            a.append(1)
            a.insert(0, 1)
            n = len(a)
            dp = [[0 for i in range(n)] for j in range(n)]
            for i in range(n-3,-1,-1):
                for j in range(i+2,n):
                    for k in range(i+1,j):
                        dp[i][j] = max(dp[i][j], dp[i][k] + dp[k][j] + a[i] * a[k] * a[j])

            return dp[0][n - 1]
